# AoC Day 4

Jenna Jordan

4 December 2021

## Prompt

--- Day 4: Giant Squid ---

You're already almost 1.5km (almost a mile) below the surface of the ocean, already so deep that you can't see any sunlight. What you can see, however, is a giant squid that has attached itself to the outside of your submarine.

Maybe it wants to play bingo?

Bingo is played on a set of boards each consisting of a 5x5 grid of numbers. Numbers are chosen at random, and the chosen number is marked on all boards on which it appears. (Numbers may not appear on all boards.) If all numbers in any row or any column of a board are marked, that board wins. (Diagonals don't count.)

The submarine has a bingo subsystem to help passengers (currently, you and the giant squid) pass the time. It automatically generates a random order in which to draw numbers and a random set of boards (your puzzle input). For example:

```
7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1

22 13 17 11  0
 8  2 23  4 24
21  9 14 16  7
 6 10  3 18  5
 1 12 20 15 19

 3 15  0  2 22
 9 18 13 17  5
19  8  7 25 23
20 11 10 24  4
14 21 16 12  6

14 21 17 24  4
10 16 15  9 19
18  8 23 26 20
22 11 13  6  5
 2  0 12  3  7
```

After the first five numbers are drawn (7, 4, 9, 5, and 11), there are no winners, but the boards are marked as follows (shown here adjacent to each other to save space):

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```
After the next six numbers are drawn (17, 23, 2, 0, 14, and 21), there are still no winners:

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

Finally, 24 is drawn:

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

At this point, the third board wins because it has at least one complete row or column of marked numbers (in this case, the entire top row is marked: 14 21 17 24 4).

## Part 1

The score of the winning board can now be calculated. Start by finding the sum of all unmarked numbers on that board; in this case, the sum is 188. Then, multiply that sum by the number that was just called when the board won, 24, to get the final score, 188 * 24 = 4512.

To guarantee victory against the giant squid, figure out which board will win first. What will your final score be if you choose that board?

In [1]:
# get input
import copy

with open("inputs/day4.txt") as file:
    num_line = file.readline()
    num_list = num_line.rstrip().split(",")
    boards_lines = file.readlines()

boards_list = []
board_counter = 0
for line in boards_lines:
    line = line.rstrip()
    if line == '':
        boards_list.append([])
        board_counter += 1
    else:
        line_numbers = line.split()
        marked_line_numbers = []
        for num in line_numbers:
            marked_line_numbers.append((num,0))
        boards_list[board_counter-1].append(marked_line_numbers)

boards_list_part2 = copy.deepcopy(boards_list)

In [2]:
def mark_bingo_board(called_number, board):
    for row_num, row in enumerate(board):
        for spot_num, spot in enumerate(row):
            num, mark = spot
            if num == called_number:
                board[row_num][spot_num] = (num,1)

In [3]:
def test_for_victory(board):
    for row in board:
        row_sum = 0
        for spot in row:
            num, mark = spot
            row_sum += mark
        if row_sum == 5:
            return True
    for c in range(len(board[0])):
        column_sum = 0
        for row in board:
            num, mark = row[c]
            column_sum += mark
        if column_sum == 5:
            return True
    
    return False       

In [4]:
def calculate_final_score(board, winning_number):
    unmarked_nums = 0
    for row in board:
        for spot in row:
            num, mark = spot
            if mark == 0:
                unmarked_nums += int(num)
    return unmarked_nums * int(winning_number)

In [5]:
victory = False
for called_num in num_list:
    if victory:
        break
    for board in boards_list:
        mark_bingo_board(called_num, board)
        victory = test_for_victory(board)
        if victory:
            final_score = calculate_final_score(board, called_num)
            break

print(final_score)

82440


### Part 1 Solution

82440 - CORRECT!

## Part 2
--- Part Two ---

On the other hand, it might be wise to try a different strategy: let the giant squid win.

You aren't sure how many bingo boards a giant squid could play at once, so rather than waste time counting its arms, the safe thing to do is to figure out which board will win last and choose that one. That way, no matter which boards it picks, it will win for sure.

In the above example, the second board is the last to win, which happens after 13 is eventually called and its middle column is completely marked. If you were to keep playing until this point, the second board would have a sum of unmarked numbers equal to 148 for a final score of 148 * 13 = 1924.

Figure out which board will win last. Once it wins, what would its final score be?

In [6]:
last_victory = False
win_tracker = {}
for called_num in num_list:
    if last_victory:
        break
    for board_num, board in enumerate(boards_list_part2):
        mark_bingo_board(called_num, board)
        victory = test_for_victory(board)
        if victory:
            win_tracker[board_num] = True
            if all(win_tracker.values()):
                last_victory = True
        else:
            win_tracker[board_num] = False
        if last_victory:
            final_score = calculate_final_score(board, called_num)
            break

print(final_score)

20774


## Part 2 Solution

20774 - CORRECT!

(on second attempt)

At first submitted:

10365 - WRONG: had a win count and did not account for boards winning again after first win